### ISSUE TITLE
Locales?

### ISSUE THREAD

**Issue by aolko** [xx45/dayjs#36]
will there be any locales avaliable?

**Comment by iamkun** [comment]
Yes. 
But I am looking for a simple implementation, not to expand our library.
e.g.
```
import dayjs from 'dayjs'

dayjs.setWeekName(['dimanche', 'lundi', 'mardi', ...]) // French

dayjs.format()// then return locales
```
How do you think?

**Comment by aolko** [comment]
sorry, i don't speak ES5/6 😅 

maybe stick to a moment.js approach for now?

**Comment by iamkun** [comment]
😅 ok, here is the compiled one in browser
```html
<script src="https://unpkg.com/dayjs"></script>
<script>
// French locales
dayjs.setWeekName(['dimanche', 'lundi', 'mardi', ...]) 
// return locales [dimanche]
dayjs().format('[dddd]') 
</script>
```

**Comment by aolko** [comment]
hmm, alright
how about "from now","ago" localisations?

**Comment by iamkun** [comment]
This is much like `moment().fromNow()` API dose.

But do we really need an API like this? 

I mean perhaps some one want to return a string "4 years", another one want "4 years ago", and another one want something else(number 4 maybe). And maybe it's hard to satisfied them all.

How do you think?

**Comment by b44rd** [comment]
How about going for something like this for handling language overrides? I think it looks pretty sweet. 

```
// Only override weekdays and months
import { weekdays, months } from 'dayjs/locale/fr'
dayjs.setLocale({ weekdays, months })
```

Alternatively, for including an entire language file

```
// Override entire language
import * as locale from 'dayjs/locale/fr'
dayjs.setLocale(locale)
```

**Comment by mauron85** [comment]
I just came here from hn. I already had proposal for date-fns to use Intl.DateTimeFormat for most case, so only can suggest same here. :) Good job so far.
https://github.com/date-fns/date-fns/issues/464

**Comment by Frondor** [comment]
I like b44rd approach. Just pass an object of key-value strings and let dayjs initialize with it, by overriding the default strings with the provided ones.

**Comment by schiem** [comment]
I'll start working on a PR for the approach outlined by @b44rd , unless someone has an objection.

**Comment by SBoudrias** [comment]
I think a big mistake moment did is making the configured locale global.

This is a big issue for Node.js server side where many requests might be processed at the same time.

Having a API returning a new dayjs instance with the localization would be better:

```js
import dayjs from 'dayjs';
import withLocale from 'dayjs/locale';
import frenchConf from 'dayjs/locale/fr-FR';

const dayjsFR = withLocale(dayjs, { weekdays: ['Dimanche', 'Lundi', ...etc] });
const dayjsFR2 = withLocale(dayjs, frenchConf);
```

**Comment by Frondor** [comment]
@SBoudrias you could store different instances per language with this approach #77
```
const dayjsFR = dayjs(null, { WD: ['Dimanche', 'Lundi', ...etc] });
```
Pinging @schiem :smiley:

**Comment by schiem** [comment]
That looks good to me!  That's really similar to the approach I was thinking of. 

I think if we add localization, we'll need to rethink how abbreviated months are done, since it currently just takes the first the first three letters of the month, which isn't in line with how other languages handle the abbreviation.  My guess is that the only fully localized way to handle that will be with a separate array of abbreviations. 

Apart from that consideration, the only other thing I can think of is adding a few more tests to make sure that the language specific set/get are working (tests for calling `dayjs(...).set('minuto', 10)`).

**Comment by Frondor** [comment]
I don't think how months are being abbreviated is a problem. I'm pretty sure the first three letters convention is world-wide accepted. I'd be worried about pluralization instead.

In the future, for newer formats, we may need some "internationalized" helper for pluralization. Since, for instance, the singular for `months` is `month` (without last `s`), but in spanish, `meses` (which means `months`) is just `mes` (without last `es`).

> Apart from that consideration, the only other thing I can think of is adding a few more tests to make sure that the language specific set/get are working (tests for calling dayjs(...).set('minuto', 10)).

In regards to that, we must discuss if `set` is using the localized word or not. If you ask me, I vote to keep it in english, so the whole source code still operates using the english conventions.
So you internationalize your `dayjs` instances, but still writing code like `dayjs(...).set('minutes', 10)` (it also helps to follow documentation examples)

**Comment by iamkun** [comment]
@Frondor Looks Nice.

I just wondering if it's better returning a new function instead of a new instance.

```
import dayjs from 'dayjs'
import localeES from 'dayjs/locales/es'

const spanishDayjs = dayjs.locale(localeES)
// then just use spanishDayjs later

spanishDayjs().set(...).format(...)

```

**Comment by schiem** [comment]
Unless I'm reading the PR incorrectly (I may be, correct me if I am), the changes at index.js:137-156 will change the `set` function to require the localized language to be used (the switch statement will go from `case 'minute':` to `case 'minuto':`, since the constant `this.$C.MIN` is being changed with the localization).

Would a pluralizer function that can be overridden in the localization work?  So for English the default function would return the input string + 's', but for other languages you could handle the edge cases directly in the function with if tests.  The trade off would be that the library would stay super small since the English function could just be one line, but the pluralizer function might be unwieldy in other languages.

**Comment by iamkun** [comment]
@Frondor Another idea here. We might better give our user a choice whether to set local in current line of code, a new instance, or set global.

A pseudo code here:
```javascript
let local = 'Monday'
class Dayjs {
  constructor(config, local) {
    this.local = null
    if (local) this.local = local
  }

  setGlobalLocal(l) {
    local = l
  }

  format() {
    return this.local || local
  }
}
const dayjs = (config, local) => (new Dayjs(config, local))

dayjs.newLocalDayjs = (local) => (config) => (new Dayjs(config, local))

export default dayjs 
```

And usage:
```javascript
// 1
dayjs() // normal dayjs
// 2
dayjs(20180808, 'spanishDay') // set local in this line only
dayjs() // still normal dayjs
// 3
dayjs.setGlobalLocal('spanishDay') // dayjs turns to spanish 
dayjs() // spanish dayjs
// 4
const FrenchDayjs = dayjs.newLocalDayjs('FrenchDay') // get a new FrenchDayjs function
FrenchDayjs() // french dayjs
dayjs() // still normal dayjs
``` 
@schiem @SBoudrias opinions?

**Comment by Frondor** [comment]
I don't think we should encourage users to set different locales per instance. Remember every time you provide locales you're doing `Object.assign` computations.
Anyway, can you think of a real world example where someone needs to provide formatted dates for different languages from the same instance, probably in the same page? If so, and it turns to be useful, then I suggest a second parameter in `dayjs(...).add(...).format('...', overrides)`. 

At this moment (in my PR), the whole cycle lets you add / switch / override locales three times: `dayjs(20180808, frFR).setLocale(enGB).format('...', esES)`

I know it's out of the scope of this issue, but I'll show an "advanced" use case for this lib; we've building a lightweight ORM and my goal is to provide date casting with `dayjs`. I can share an example of how it works with localization.
But for this to work, I want to suggest a change on the way how `dayjs` actually instantiates itself. So a default `format` can be provided for `JSON.stringify(dayjsObject)` to use through (the missing) `dayjs.toJSON` method.

You may ask _why not `dayjs().defaultFormat('...')`?_, well... In my humble opinion, I consider the method chain is actually pretty long and with many configurations incoming (as the project grows), you'll see your how code ends up looking better as `dayjs(initConfig)` instead of `dayjs().setXconfig().setYconfig().setZconfig()...`. **But once again, it's my opinion, and the use case I'm showing as example is just to help me explain my point**.

We can easily solve that by doing something like this:
> The class is exported for the user to manually instantiate; let's say `import dayjs from 'dayjs'` imports an instance, and `import { Dayjs } from 'dayjs'` imports the class.
```js
import { Dayjs }, dayjs from 'dayjs'

const FrenchDayjs= Dayjs({ format: 'YYYY', locale: frFR, date: nullValue || new Date() })
// or
const date = dayjs(new Date())

console.log(JSON.stringify(Dayjs)) // prints "2018" because it internally uses obj.toJSON() which uses default format provided
```

You can see how this kind of use case benefits from the new config here (note the `MockDayjs` object):
https://codepen.io/Frondor/pen/BxpLZG?editors=0012
(It's tedious, but easy to follow the way how I use dayjs to provide date casting)
```js
const car = new Car({ created_at: new Date(2018, 4, 28) }) // car.created_at === Dayjs object
car.created_at.format('YYYY') // outputs "2018"
JSON.stringify(car) // Object: { created_at: "Saturday of 2018" }
```

**Comment by iamkun** [comment]
Still, I'm afraid I didn't get what you mean. Sorry. May be I have to go through your code and comment again tomorrow.

Questions, 
- 1 why we are going to change dayjs#toJSON  ?
- 2 what's this ORM here stand for? (Confused me most)

Plus, any direct opinion or suggestion to my pseudo code above? Looking for a feed back. :D 

Good job so far. I think we are close to the FINAL locales we are looking for.

**Comment by Frondor** [comment]
> Plus, any direct opinion or suggestion to my pseudo code above? Looking for a feed back. :D

Looks similar to my approach, but the `newLocalDayjs` approach adds a bit of unnecessary complexity in my humble opinion. Also, if you pass a string to `dayjs(20180808, 'spanishDay')` , that means you'll be requiring locale objects dynamically, and the UMD version is going to ship with every locale we support, that's going to increase the size a lot and even moment.js doesn't do that. Or probably you meant an object and not a string? Anyway, that's why I suggest to simply
```js
export const class Dayjs { constructor (config) {...} }
export default const dayjs = (date, config) => new Dayjs(config) } // config.date = date
```
and let the user handle the instance however he wants (see my `MockDayjs` example)

**Comment by iamkun** [comment]
>  and the UMD version is going to ship with every locale we support

My mistake. It's an imported object not a string. I've corrected the code above.

> The class is exported for the user to manually

Export `Dayjs` class is a good idea. Cause this ships more flexibility.

**Comment by Frondor** [comment]
I agree with both ideas. At the moment, we don't have `dayjs.local` but [`dayjs.setLocale`](https://github.com/Frondor/dayjs/blob/4242bbf7c3cb070c0870252fc165ab89216480f2/test/localization.test.js#L25) and [`dayjs.format(..., { locale: FR })`](https://github.com/Frondor/dayjs/blob/4242bbf7c3cb070c0870252fc165ab89216480f2/test/localization.test.js#L30). It's a matter of taste I think, I don't really mind on which term to use for this method.

You can also do [`dayjs(new Date(), { locale: FR })`](https://github.com/Frondor/dayjs/blob/4242bbf7c3cb070c0870252fc165ab89216480f2/test/localization.test.js#L14)

> or just one api for less confusion. In this way, a locale is also a kind of plugin:

Now that I think, I prefer your first example with `dayjs.extend((instance) => instance.$xyz = 123)`, it's super easy to work with and all it does is to "pipe" a callback with access to the dayjs instance for the user to mutate it.

**Comment by iamkun** [comment]
OK, here comes a problem. How can we set a global locale config?
It might because of the limitation of the structure in our current code.
We can do `dayjs().setLocale(esES)` on a instance, but we can't do `dayjs.setLocale(esES)` globally. Isn't it?

Any suggestion about a global configuration?
```javascript
dayjs.local(esES)
dayjs.format() // esES by default
dayjs.format() // esES
```

**Comment by iamkun** [comment]
Plus, @Frondor   @schiem  would you mind making a new pull request to branch `@next`, instead of direct to master.

We are currently implementing two important features locales as well as plugin system. And our code might dependence others.

I want get everything(test, docs, rollup config...) done before merge to our main branch. 

Thanks

**Comment by schiem** [comment]
@iamkun Absolutely.  Would you prefer a new PR or just modifying the existing one?

**Comment by Frondor** [comment]
Alright, we can't change the target branch, so I have to close this PR and send a new one to the new branch, it actually targets `dev`.

Anyway, what's the plugin system? Can't we tackle it by just creating an `extend` method that receives a callback with `this` passed to it? i.e.: `extend (fn) { fn(this); return this }` or even `extend (fn) { fn.call(this); return this }`

> OK, here comes a problem. How can we set a global locale config?

Actually, I think there's not way we can do that without using globals or env variables (I recommend none of both). I suggest to encourage the user to create base instances with its own config, and then `set()` / `clone()` it

I've written a full example of what's done so far in my PR, #77 and new approachs to some methods

**Comment by schiem** [comment]
@Frondor The original intent behind the plugins was to also allow overwriting built in methods, e.g. allowing more format options without needing to add bloat to the core.  So you could still call `dayjs().format("some arcane format option")` instead of `dayjs().pluginFormat("...")`.  And since it takes the place of the original method, it has access to `this` by default.  I'm not 100% sure that this is the best approach, I'm planning on giving it another look when I get home to see if I can come up with something more elegant.

> OK, here comes a problem. How can we set a global locale config?

The way that I handled that for the plugin system was by having an instance set it on the prototype, like:
```
function extend(plugin) {
    const proto = Object.getPrototypeOf(this)
    proto[plugin.name] = plugin.method
}
```
Could overwriting a property of the prototype be done to set locales?

**Comment by Frondor** [comment]
@schiem you can take a look to the [Vue](https://vuejs.org/v2/guide/plugins.html) approach to plugins, you may extract some ideas from there. I personally like how the install method just receives the `Vue` object and you add / overrides to its prototype. Then it does some work internally depending on what you registered (for example, merging the mixins and creating the directive objects, but that's not really needed here)

> Could overwriting a property of the prototype be done to set locales?

Can't see the necessity for that though, the object locales is always living in the `$L` property. Please, see how it works here #90 Every review is welcome
